python/ray/worker.py

```diff
@@ -854,21 +850,79 @@ def _handle_process_task_failure(self, function_descriptor,
         # Send signal with the error.
         ray_signal.send(ray_signal.ErrorSignal(str(failure_object)))
 
-    def _wait_for_and_process_task(self, task):
+    def _process_normal_task(self, function_descriptor_list, job_id, task_id,
+                             arguments, return_ids):
         """Wait for a task to be ready and process the task.
 
         Args:
             task: The task to execute.
         """
+        # Automatically restrict the GPUs available to this task.
+        ray.utils.set_cuda_visible_devices(ray.get_gpu_ids())
+
         function_descriptor = FunctionDescriptor.from_bytes_list(
-            task.function_descriptor_list())
-        job_id = task.job_id()
+            function_descriptor_list)
+
+        execution_info = self.function_actor_manager.get_execution_info(
+            job_id, function_descriptor)
+
+        # Execute the task.
+        function_name = execution_info.function_name
+        extra_data = {"name": function_name, "task_id": task_id.hex()}
+        title = "ray_worker:{}()".format(function_name)
+        next_title = "ray_worker"
+
+        with profiling.profile("task", extra_data=extra_data):
+            with _changeproctitle(title, next_title):
+                function_executor = execution_info.function
+                function_name = execution_info.function_name
+
+                self.reraise_actor_init_error()
+                self.memory_monitor.raise_if_low_memory()
```

Comment:
These should be in the try/except block, otherwise the error is ignored and the worker hangs. Same for process_actor_task below.

Also, it seems the core worker doesn't handle any unexpected errors in the execution callback gracefully.